### PR TITLE
TECH-122: Generic mockoon api added

### PR DESCRIPTION
**Pull request by dborowiecki**
Ticket: [TECH-122](https://govstack-global.atlassian.net/browse/TECH-122)
## Description
This contribution provides proof of concept for generic Mockoon API example created from OpenAPI schema. 
Dockerized application gathers all .yaml files from /api subdirectory, and creates single schema used by mockoon-cli. Pasting example/mockoon into other building blocks should work without additional configuration changes.

### CODE REVIEW

**Review by KarolinaKopacz — APPROVED** [review]

File: examples/mockoon/api_organizer.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// Script used in mockoon API assembly. Executed in docker-compose it merges all api files into single 
+// file with common name. Should be executed from root
+
+const path = require('path');
+const yaml = require('js-yaml');
+const YAML = require('yaml')
+const fs = require('fs');
+
+const directoryPath = path.join(__dirname, 'api');
+const assemblyApiFileName = 'AssemblyApiDefinition.yaml'
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:10]:
```suggestion
const assemblyApiFileName = 'AssemblyApiDefinition.yaml';
```

File: examples/mockoon/api_organizer.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+function validateUnique(existingPaths, newPaths, propertyType) { 
+    for (const property in newPaths) {
+        if (property in existingPaths){
+            console.warn(`Assembly's ${propertyType} ${property} already defined in assembly API. Definition will be overwritten.`);
+        }
+    }
+}
+
+function extendObjectProperty(existingObj, newObj, property) {
+    validateUnique(existingObj[property], newObj[property], property)
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:21]:
```suggestion
    validateUnique(existingObj[property], newObj[property], property);
```

File: examples/mockoon/api_organizer.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    if (!!newObj[property]) {
+        existingObj[property] = {...existingObj[property], ...newObj[property]}
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:23]:
```suggestion
        existingObj[property] = {...existingObj[property], ...newObj[property]};
```

File: examples/mockoon/api_organizer.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+}
+
+function saveApiDefinition(jsonApiDefinition) {
+    const filePath = path.join(directoryPath, assemblyApiFileName)
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:28]:
```suggestion
    const filePath = path.join(directoryPath, assemblyApiFileName);
```

File: examples/mockoon/api_organizer.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    const doc = new YAML.Document();
+    doc.contents = jsonApiDefinition;
+    fs.writeFile(filePath, doc.toString(), function (err) {
+        if (err) return console.error(err);
+        console.log(`YAML definition saved as ${filePath}`);
+    });
+}
+
+function loadApiYAMLs() {
+    let fileArray = []
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:38]:
```suggestion
    let fileArray = [];
```

File: examples/mockoon/api_organizer.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    doc.contents = jsonApiDefinition;
+    fs.writeFile(filePath, doc.toString(), function (err) {
+        if (err) return console.error(err);
+        console.log(`YAML definition saved as ${filePath}`);
+    });
+}
+
+function loadApiYAMLs() {
+    let fileArray = [] 
+    let assembly =    {
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:39]:
```suggestion
    let assembly = {
```

File: examples/mockoon/api_organizer.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    fs.readdir(directoryPath, function (err, files) {
+        //handling error
+        if (err) {
+            return console.log('Unable to scan directory: ' + err);
+        } 
+        //listing all files using forEach
+        files.forEach(function (file) {
+            // Do whatever you want to do with the file
+            if (file.endsWith('yaml')) {
+                const apiDir = path.join(directoryPath, file)
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:66]:
```suggestion
                const apiDir = path.join(directoryPath, file);
```

File: examples/mockoon/api_organizer.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                const nextYAML = yaml.load(fs.readFileSync(apiDir, 'utf8'))
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:67]:
```suggestion
                const nextYAML = yaml.load(fs.readFileSync(apiDir, 'utf8'));
```

File: examples/mockoon/api_organizer.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                extendObjectProperty(assembly, nextYAML, 'paths')
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:68]:
```suggestion
                extendObjectProperty(assembly, nextYAML, 'paths');
```

File: examples/mockoon/api_organizer.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                extendObjectProperty(assembly.components, nextYAML.components, 'schemas')
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:69]:
```suggestion
                extendObjectProperty(assembly.components, nextYAML.components, 'schemas');
```

File: examples/mockoon/api_organizer.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                extendObjectProperty(assembly.components, nextYAML.components, 'parameters')
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:70]:
```suggestion
                extendObjectProperty(assembly.components, nextYAML.components, 'parameters');
```

File: examples/mockoon/api_organizer.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+                extendObjectProperty(assembly.components, nextYAML.components, 'securitySchemes')
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:71]:
```suggestion
                extendObjectProperty(assembly.components, nextYAML.components, 'securitySchemes');
```

File: examples/mockoon/api_organizer.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            } else {
+                console.warn("Found non YAML file in api directory. It'll not used in API definiion. [%s]", file)
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:73]:
```suggestion
                console.warn("Found non YAML file in API directory. It'll not used in API definition. [%s]", file);
```

File: examples/mockoon/api_organizer.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+            }
+        });
+        saveApiDefinition(assembly)
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:76]:
```suggestion
        saveApiDefinition(assembly);
```

File: examples/mockoon/api_organizer.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    });
+}
+
+loadApiYAMLs()
```
Comment (dszafranek) [inline examples/mockoon/api_organizer.js:80]:
```suggestion
loadApiYAMLs();
```

**Review by dszafranek — CHANGES_REQUESTED** [review]

**Comment by dborowiecki** [comment]
@dszafranek Thanks for detailed review. There were multiple issues related to styles, to avoid that in the future I've attached eslint to the project.

**Review by dszafranek — APPROVED** [review]

**Pull request merged**